primitives/consensus/common/src/block_import.rs

```
@@ -223,6 +224,32 @@ impl<Block: BlockT, Transaction> BlockImportParams<Block, Transaction> {
 			import_existing: self.import_existing,
 		}
 	}
+
+	/// Take interemdiate by given key, and remove it from the processing list.
+	pub fn take_intermediate<T: 'static>(&mut self, key: &'static [u8]) -> Option<Box<T>> {
```

For all these functions I would return an error when the downcasting fails.